wnfs/src/common/utils.rs

```diff
@@ -66,6 +67,29 @@ pub(crate) fn split_last(path_segments: &[String]) -> Result<(&[String], &String
     }
 }
 
+pub(crate) async fn read_fully(
+    stream: &mut (impl AsyncRead + Unpin),
+    buffer: &mut [u8],
+) -> Result<(usize, bool)> {
+    let mut bytes_read = 0;
+    let mut done = false;
+    loop {
+        let bytes_read_in_iteration = stream.read(&mut buffer[bytes_read..]).await?;
+
+        bytes_read += bytes_read_in_iteration;
+
+        if bytes_read_in_iteration == 0 {
+            done = true;
+            break;
+        }
+
+        if bytes_read == buffer.len() {
+            break;
+        }
+    }
+    Ok((bytes_read, done))
+}
+
```

Comment:
Interesting, `read_to_end` won't be sufficient because it can resize the input buffer.

Reply:
Yeah exactly. I wanted something with fewer copies.
I know the buffer size in advance (it's the file chunk size) and just fill it.
`read_to_end` just gives me *some* size, and I don't have any control over it.